forgerock-authenticator/src/main/java/org/forgerock/android/authenticator/PushResponder.java

```diff
@@ -100,22 +109,72 @@ public static int respond(String endpoint, String amlbCookie, String base64Secre
+            url = new URL(endpoint);
+            connection = (HttpURLConnection) url.openConnection();
+
+            // Set request parameters and establish connection
+            connection.setDoOutput(true);
+            connection.setDoInput(true);
+            connection.setRequestMethod("POST");
+            connection.setRequestProperty("Content-Type", "application/json");
+            connection.setRequestProperty("Accept-API-Version", "resource=1.0, protocol=1.0");
+            if (amlbCookie != null) {
+                connection.setRequestProperty("Cookie", amlbCookie);
+            }
+        } catch (MalformedURLException e) {
+            Logger.warn(TAG, e, "Error establishing connection. URL malformed or invalid.");
+            throw new IOException("Invalid URL!", e);
+        } catch (ProtocolException e) {
+            Logger.warn(TAG, e,"Error establishing connection. POST requested method " +
+                    "isn't valid for the endpoint.");
+            throw new IOException("Invalid URL!", e);
+        }
+
+        return connection;
+    }
+
+    private static String generateJwt(String base64Secret, Map<String, Object> data)
+            throws IOException, JOSEException {
+
+        // Check shared secret
+        if(base64Secret == null || base64Secret.length() == 0) {
+            Logger.debug(TAG, "Error generating JWT data. Secret is empty or null.");
+            throw new IOException("Passed empty secret");
+        }
+
+        // Prepare JWT with claims
+        JWTClaimsSet.Builder claimBuilder = new JWTClaimsSet.Builder();
         for (String key : data.keySet()) {
-            builder.claim(key, data.get(key));
+            claimBuilder.claim(key, data.get(key));
         }
 
-        byte[] secret = Base64.decode(base64Secret);
+        // Apply the HMAC protection
+        JWSHeader header = new JWSHeader.Builder(JWSAlgorithm.HS256)
+                .type(JOSEObjectType.JWT)
+                .build();
+
+        SignedJWT signedJWT = new SignedJWT(header, claimBuilder.build());
 
-        SigningHandler signingHandler = new SigningManager().newHmacSigningHandler(secret);
-        SignedJwtBuilderImpl jwtBuilder = new SignedJwtBuilderImpl(signingHandler);
-        jwtBuilder.claims(builder.build());
-        jwtBuilder.headers().alg(JwsAlgorithm.HS256);
         try {
-            return jwtBuilder.build();
-        } catch (IllegalArgumentException e) {
-            throw new IOException("Passed empty secret", e);
+            // Create HMAC signer
+            byte[] secret = Base64.decode(base64Secret, Base64.NO_WRAP);
+            JWSSigner signer = new MACSigner(secret);
+
+            // Sign JWT
+            signedJWT.sign(signer);
+            return signedJWT.serialize();
+        } catch (IllegalArgumentException | KeyLengthException e) {
+            Logger.debug(TAG, "Error generating JWT data. Secret malformed or invalid.");
+            throw new IOException("Invalid secret!", e);
```

Comment:
Yes, IllegalArgumentException and KeyLengthException does not extends from IOException. However, what cause those exceptions are the issue with the input data, the reason why we are using a IOException. It can be changed to throw the original exception, but is there any problem it can cause if not changed?

Reply:
No issue, we can always wrap one exception to another, just want to make it more meaningful and with good practice, when caller receives IOException, it would expect something related to IO.